core/jobs/batch_jobs/audit_rule_validation_jobs.py

```diff
@@ -0,0 +1,2194 @@
+            ):
+                invalid_states.append(
+                    {
+                        'state_name': state_name,
+                        'ans_group_idx': invalid_ans_groups
+                    }
+                )
+
+            choices_less_than_4 = len(choices) - len(invalid_choices_index) < 4
+            if choices_less_than_4:
+                invalid_states_for_choices_less_than_4.append(state_name)
+
+        return (invalid_states, invalid_states_for_choices_less_than_4)
+
+    @staticmethod
+    def text_input_invalid_values(
+        states_dict: Dict[str, state_domain.State]
+    ) -> List[Dict[str, object]]:
+        """Checks the following in the MultipleChoiceInput interaction
+            - `Contains` should always come after `Equals` and `Startswith`
+            rules
+            - `Startswith` should always come after the `Equals` rule
+            - Rule is duplicate
+
+        Args:
+            states_dict: dict[str, state_domain.State]. The state dictionary.
+
+        Returns:
+            invalid_states: List[Dict[str, object]]. List of invalid states
+            with the invalid answer groups.
+        """
+        invalid_states = []
+        for state_name, state in states_dict.items():
+            if state.interaction.id != 'TextInput':
+                continue
+            seen_strings_contains = []
+            seen_strings_startswith = []
+            rule_spec_till_now = []
+            invalid_ans_groups = []
+            valid_dest_ans_group_count = 0
+            invalid_ans_group_count = 0
+            answer_groups = state.interaction.answer_groups
+            for ans_group_idx, answer_group in enumerate(answer_groups):
+                invalid_rules = []
+                for rule_spec in answer_group.rule_specs:
+                    # Rule is duplicate.
+                    if rule_spec.to_dict() in rule_spec_till_now:
+                        invalid_rules.append(rule_spec)
+                    else:
+                        rule_spec_till_now.append(rule_spec.to_dict())
+
+                    if rule_spec.rule_type == 'Contains':
+                        seen_strings_contains.append(
+                            rule_spec.inputs['x']['normalizedStrSet'])
+
+                    if rule_spec.rule_type == 'StartsWith':
+                        seen_strings_startswith.append(
+                            rule_spec.inputs['x']['normalizedStrSet']
+                        )
+                        # `Contains` should always come after `Equals` and
+                        # `Startswith` rules.
+                        rule_values = rule_spec.inputs['x']['normalizedStrSet']
+                        for contain_ele in seen_strings_contains:
+                            for item in contain_ele:
+                                for ele in rule_values:
+                                    if item in ele:
+                                        invalid_rules.append(rule_spec)
```

Comment:
Can you be more descriptive in naming these variables?

Reply:
Done, please let me know if it looks good to you.